From: Hugh Dickins <hughd@google.com>
Date: Thu, 13 Jan 2011 15:46:28 -0800
Subject: [PATCH] thp: ksm: free swap when swapcache page is replaced
References: THP core (fate #311931)
Git-commit: ae52a2adb5afa5ac5ec5fb5c7b24777f84b6c926
Patch-mainline: v2.6.38-rc1

When a swapcache page is replaced by a ksm page, it's best to free that
swap immediately.

Reported-by: Andrea Arcangeli <aarcange@redhat.com>
Signed-off-by: Hugh Dickins <hughd@google.com>
Signed-off-by: Andrea Arcangeli <aarcange@redhat.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
Signed-off-by: Mel Gorman <mgorman@suse.de>
---
 mm/ksm.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/mm/ksm.c b/mm/ksm.c
index 164eed5..ceb0c59 100644
--- a/mm/ksm.c
+++ b/mm/ksm.c
@@ -812,6 +812,8 @@ static int replace_page(struct vm_area_struct *vma, struct page *page,
 	set_pte_at_notify(mm, addr, ptep, mk_pte(kpage, vma->vm_page_prot));
 
 	page_remove_rmap(page);
+	if (!page_mapped(page))
+		try_to_free_swap(page);
 	put_page(page);
 
 	pte_unmap_unlock(ptep, ptl);

